File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.cpp
class Solution {
private:
    int gcd(int n1, int n2){
        while (n2 > 0){
            n1 %= n2;
            swap(n1, n2);
        }
        return n1;
    }

public:
    string findLexSmallestString(string s, int a, int b) {
        string t, ans = s;
        int n = s.size(), g = gcd(n, b);
        s.append(s);

        auto add = [&](string& t, int start){
            int minVal = 10, times = 0;
            for (int i = 0; i < 10; ++i){
                int added = ((t[start] - '0') + i * a) % 10;
                if (added < minVal){
                    minVal = added;
                    times = i;
                }
            }

            for (int i = start; i < n; i += 2) t[i] = '0' + ((t[i] - '0') + times * a) % 10;
        };

        for (int i = 0; i < n; i += g){
            t = s.substr(i, n);
            add(t, 1);

            if (b % 2) add(t, 0);
            ans = min(ans, t);
        }

        return ans;
    }
};

File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.py
class Solution:
    def findLexSmallestString(self, s: str, a: int, b: int) -> str:
        def gcd(a: int, b: int) -> int:
            while b > 0:
                a %= b
                a, b = b, a
            return a
        
        def add(t: str, start: int) -> str:
            original = int(t[start])
            min_val, times = 10, 0

            for i in range(10):
                added = (original + i * a) % 10
                if added < min_val:
                    min_val = added
                    times = i
            
            t_list = list(t)
            for i in range(start, n, 2):
                t_list[i] = str((int(t_list[i]) + times * a) % 10)
            return "".join(t_list)

        n, ans = len(s), s
        s += s
        g = gcd(n, b)

        for i in range(0, n, g):
            t = s[i : i + n]
            t = add(t, 1)
            if b % 2:
                t = add(t, 0)
            
            if t < ans:
                ans = t
        
        return ans
